File: lamquant-firmware/src/stack_guard.rs
//! Hardware stack overflow trap via RISC-V Physical Memory Protection (PMP).
//!
//! Port of `firmware/core/stack_guard.c`. The Hazard3 core implements PMP
//! per the RISC-V Privileged Architecture spec. We configure PMP entry 0
//! as a NAPOT region covering 2 KB at the stack-top guard zone. Locked
//! at boot (`L = 1`) so it cannot be modified afterward — any access to
//! that region traps and we branch to safe mode.

const STACK_GUARD_ADDR: u32 = 0x2000_7800;
const STACK_GUARD_SIZE: u32 = 2048; // 2 KB

// pmpcfg encoding:
const PMP_CFG_L: u32 = 1 << 7; // Lock bit — entry immutable until reset
const PMP_CFG_A_NAPOT: u32 = 3 << 3; // Naturally-Aligned Power-Of-Two
                                     // No R/W/X bits set = poisonous region

/// Install the PMP-locked stack guard. Call exactly once, before any
/// task could underflow the stack into the guard region.
///
/// After this returns, the 2 KB region at `STACK_GUARD_ADDR` is hardware-
/// protected. Any load/store there raises a precise access fault trap.
#[inline(never)]
pub fn install() {
    // pmpaddr0 = (base >> 2) | ((size / 8) - 1)
    // For NAPOT: low bits encode the region size as a string of 1s.
    // 2 KB / 8 - 1 = 0xFF (8 trailing 1-bits).
    let pmpaddr0: u32 = (STACK_GUARD_ADDR >> 2) | ((STACK_GUARD_SIZE / 8) - 1);
    let pmpcfg0: u32 = PMP_CFG_L | PMP_CFG_A_NAPOT;

    // SAFETY: csrw to pmpaddr0/pmpcfg0 is safe at boot. The locked PMP
    // entry only restricts access to the guard zone; it cannot make
    // existing valid memory inaccessible.
    unsafe {
        core::arch::asm!(
            "csrw pmpaddr0, {addr}",
            "csrw pmpcfg0, {cfg}",
            addr = in(reg) pmpaddr0,
            cfg = in(reg) pmpcfg0,
            options(nostack, preserves_flags),
        );
    }
}

File: reference_implementations/c_firmware/core/stack_guard.c
#include <stdint.h>
#include <stdbool.h>
#include "power_states.h"
#include "stack_guard.h"

/**
 * Hardware Stack Enforcer - LamQuant Gen 6
 * RP2350 Physical Memory Protection (PMP) Core Bindings
 */

#define PMP_CFG_L         (1 << 7)    // Lock bit protecting state registers
#define PMP_CFG_A_NAPOT   (3 << 3)    // Naturally Aligned Power-Of-Two

// Injected directly via `main()` early boot initializing strict trap registers
void stack_setup_hardware_trap(void) {
    // Use NAPOT mode to protect exactly 2KB at 0x20007800
    // Address encoding: (0x20007800 >> 2) | ((2048 / 8) - 1)
    uint32_t pmpaddr0 = (0x20007800 >> 2) | 0xFF;

    // Config: Lock (L) + NAPOT (A_NAPOT). No R/W/X permissions = Poisonous range.
    uint32_t pmpcfg = PMP_CFG_L | PMP_CFG_A_NAPOT;

    __asm__ volatile (
        "csrw pmpaddr0, %0\n"
        "csrw pmpcfg0, %1\n"
        :
        : "r"(pmpaddr0), "r"(pmpcfg)
    );
}

// Exception hook implicitly branched when the pointer violates limits
void __attribute__((naked)) stack_overflow_exception(void) {
    // Immediate safe mode routing dropping allocated execution frames organically
    __asm__ volatile (
        "li   sp, 0x20008000\n"  // Reset SP to absolute TCM top
        "j    enter_safe_mode\n"
    );
}

// PMP trap is hardware-locked (L bit = 1), cannot be modified after boot.
// This is intentionally a no-op. The PMP-based stack guard at 0x20007800
// is permanently active once stack_setup_hardware_trap() runs at boot.
void stack_reset_canaries(void) {
    // No-op: PMP entries with the Lock bit are immutable until reset.
}
